pkgs/tools/misc/loop/default.nix

```diff
@@ -0,0 +1,24 @@
+{ stdenv, fetchFromGitHub, rustPlatform }:
+
+rustPlatform.buildRustPackage rec {
+  name = "loop-${version}";
+  version = "d6ef3c5a0ecd4f533908abee5e481419a1a6eeae";
+
+  src = fetchFromGitHub {
+    owner = "Miserlou";
+    repo  = "Loop";
+    rev   = version;
+    sha256 = "1fhihm32v77rj6r3scwmnvzsivky50g7a1644qrn8pafpjs4zwx5";
+  };
+
+  cargoSha256 = "1ccf8dkswwdbwf9diy0l4vc4i2g05ynhi3w1jg3b2ldrvj0j9m9s";
+
+  cargoPatches = [ ./fix_cargo_toml.patch ./fix_cargo_lock.patch ]; # Cargo.lock and Cargo.toml are not aligned
```

Comment:
Why is it needed to update _both_ Cargo.toml and Cargo.lock? Wouldn't one be enough?

Is there any reason this fix can't be submitted upstream? It is preferred to avoid custom-patching on distro level as much as possible. (Every deviation from the upstream makes it harder for the upstream author to reproduce and fix user issues.)

Reply:
I have just learned about `cargo`, so I expect to be wrong.
My understanding is that `Cargo.toml` is the file one writes with a list of direct dependencies, and `Cargo.lock` is the file generated by `cargo` with the resolved list of direct and indirect dependencies.

The problem here is that `Cargo.lock` and `Cargo.toml` disagree on 2 levels:
- dependencies version: `Cargo.lock` has more recent versions for `structopt`, `atty`, `regex`, `subprocess` and `tempfile`;
- self (= `loop`) version: `Cargo.lock` has an older version of `loop-rs`, which I assume to be `loop` itself.

My feeling is that:
- `Cargo.lock` should be trusted for dependencies: it got generated by `cargo` and witnesses a set of packages that are known to build together;
- `Cargo.toml` should be trusted for `loop` itself: this is the manifest file so the advertised version should be the reference.
Hence the double update.

Regarding an upstream patch, I didn't know whether the consistency requirement between `Cargo.toml` and `Cargo.lock` is specific to Nix, or a widespread standard, which is why I didn't dare submit a patch upstream. Should I ?

Reply:
`Cargo.lock` is the only one required to be updated.

The issue is reproducible with `cargo build --frozen`, which `buildRustPackage` uses to ensure reproducibility. Though, the issues is not Nix-specific, and it is desirable to submit the updated lock file upstream. (We could merge the patch into Nixpkgs and include the link to upstream PR, so we don't block this PR.)

FWIW, `Cargo.toml` diff is a noop—if dependency version does not have any operator in it, implicit caret operator is applied: https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio. That's why `Cargo.toml` is not updated besides `loop-rs` version.